src/core/net/geoipmanager.cpp

```diff
@@ -0,0 +1,460 @@
+static const int CACHE_SIZE = 1000;
+static const int UPDATE_INTERVAL = 30; // Days between database updates
+
+using namespace Net;
+
+// GeoIPManager
+
+GeoIPManager *GeoIPManager::m_instance = 0;
+
+GeoIPManager::GeoIPManager()
+    : m_enabled(false)
+    , m_geoIPDatabase(0)
+    , m_cache(CACHE_SIZE)
+{
+    configure();
+    connect(Preferences::instance(), SIGNAL(changed()), SLOT(configure()));
+}
+
+GeoIPManager::~GeoIPManager()
+{
+    if (m_geoIPDatabase)
+        delete m_geoIPDatabase;
+}
+
+void GeoIPManager::initInstance()
+{
+    if (!m_instance)
+        m_instance = new GeoIPManager;
+}
+
+void GeoIPManager::freeInstance()
+{
+    if (m_instance) {
+        delete m_instance;
+        m_instance = 0;
+    }
+}
+
+GeoIPManager *GeoIPManager::instance()
+{
+    return m_instance;
+}
+
+void GeoIPManager::loadDatabase()
+{
+    if (m_geoIPDatabase) {
+        delete m_geoIPDatabase;
+        m_geoIPDatabase = 0;
+    }
+
+    QString filepath = Utils::Fs::expandPathAbs(
+                QString("%1%2/%3").arg(Utils::Fs::QDesktopServicesDataLocation())
+                .arg(GEOIP_FOLDER).arg(GEOIP_FILENAME));
+
+    QString error;
+    m_geoIPDatabase = GeoIPDatabase::load(filepath, error);
+    if (m_geoIPDatabase)
+        Logger::instance()->addMessage(tr("GeoIP database loaded. Type: %1. Build time: %2.")
+                                       .arg(m_geoIPDatabase->type()).arg(m_geoIPDatabase->buildEpoch().toString()),
+                                       Log::INFO);
+    else
+        Logger::instance()->addMessage(tr("Could not load GeoIP database. %1").arg(error), Log::WARNING);
+
+    manageDatabaseUpdate();
+}
+
+void GeoIPManager::manageDatabaseUpdate()
+{
+    if (!m_geoIPDatabase || (m_geoIPDatabase->buildEpoch().daysTo(QDateTime::currentDateTimeUtc()) >= UPDATE_INTERVAL))
+        downloadDatabaseFile();
+}
+
+void GeoIPManager::downloadDatabaseFile()
+{
+    DownloadHandler *handler = DownloadManager::instance()->downloadUrl(DATABASE_URL);
+    connect(handler, SIGNAL(downloadFinished(QString, QByteArray)), SLOT(downloadFinished(QString, QByteArray)));
+    connect(handler, SIGNAL(downloadFailed(QString, QString)), SLOT(downloadFailed(QString, QString)));
+}
+
+QString GeoIPManager::lookup(const QHostAddress &hostAddr) const
+{
+    if (m_enabled && m_geoIPDatabase) {
+        if (m_cache.contains(hostAddr))
+            return *m_cache[hostAddr];
+
+        QString code = m_geoIPDatabase->lookup(hostAddr);
+        m_cache.insert(hostAddr, new QString(code));
+        return code;
+    }
+
+    return QLatin1String(ISO_CODE_NA);
+}
+
+// http://www.iso.org/iso/country_codes/iso_3166_code_lists/english_country_names_and_code_elements.htm
+QString GeoIPManager::CountryName(const QString &countryISOCode)
+{
+    static QHash<QString, QString> countries;
+    static bool initialized = false;
+
+    if (!initialized) {
+        countries[ISO_CODE_NA] = "N/A";
+        countries["AP"] = "Asia/Pacific Region";
```

Comment:
The original code didn't allow this. Any improvements on this should be postponed. First priority is to transition to the new db while doing the absolutely minimal changes to the functionality of the code. After that we can improve on anything.